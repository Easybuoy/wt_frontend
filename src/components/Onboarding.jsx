import React, { useState, useEffect } from "react";
import { withApollo } from "react-apollo";
import { Button, RadioButtonGroup, useToast } from "@chakra-ui/core";
import { useFormik } from "formik";
import * as yup from "yup";
import styled from "styled-components";

import Preview from "./common/Preview";
import AuthStyle from "./auth/AuthStyle";
import Select from "./common/Select";
import Logo from "./common/Logo";
import { ONBOARDING } from "../graphql/mutations";
import { GET_UNIT } from "../graphql/queries";
import { getUserDetails } from "../utils";

const userData = getUserDetails();

const OnboardingStyled = styled.div`
  .section-right {
    margin: auto;
    width: 50vw;
    @media only screen and (max-width: 650px) {
      width: 90%;
    }
    div {
      margin: auto;
      .btnGroup {
        .unitButton {
          border: 2px solid;
          background-color: transparent;
        }
        margin-top: 10px;
        margin-bottom: 15px;
        .unitButton:not(:last-child) {
          margin-right: 20px;
        }
      }
      .heading {
        margin-bottom: 25px;
      }
      p {
        margin-bottom: 10px;
        font-family: Roboto;
      }
      .dropdown,
      .submit {
        width: 100%;
        justify-content: space-between;
        margin-bottom: 15px;
      }
      .submit {
        background-color: #ff8744;
      }
      .dropdown {
        background: #fffcf2;
        border: 1px solid #252422;
        box-sizing: border-box;
      }
      .dropdownOptions {
        width: 30%;
      }
    }
  }
  .section-left {
    @media only screen and (max-width: 650px) {
      display: none;
    }
    width: 50vw;
    img {
      width: 100%;
      height: 100vh;
      object-fit: cover;
    }
  }
`;

const emptyAnswers = {
  heightUnit: "",
  weightUnit: "",
  goal: "",
  experience: "",
  equipment: ""
};

function Onboarding({ client, history }) {
  const toast = useToast();

  const formik = useFormik({
    initialValues: {
      goal: "",
      equipment: "",
      experience: "",
      heightUnit: "",
      weightUnit: ""
    },
    validationSchema: yup.object().shape({
      goal: yup.string().required("Please select your workout goal"),
      equipment: yup.string().required("Please enter your workout equipment"),
      experience: yup.string().required("Please enter your workout experience")
    }),

    onSubmit: value => {
      client
        .mutate({
          mutation: ONBOARDING,
          variables: {
            email: value.email,
            password: value.password,
            remember: value.remember
          }
        })
        .then(response => {
          const { token, isNewUser } = response.data.authForm;
          localStorage.setItem(
            "userData",
            JSON.stringify({ token, isNewUser })
          );
          if (isNewUser === true) {
            // setLoginSuccess("/onboarding");

            toast({
              title: "Login Successful.",
              description: "You can now complete the onboarding process",
              status: "success",
              duration: 9000,
              isClosable: true
            });
          } else {
            // setLoginSuccess("/");
            toast({
              title: "Login Successful.",
              description: "You can now access your dashboard",
              status: "success",
              duration: 9000,
              isClosable: true
            });
          }
        })
        .catch(error => {
          toast({
            title: "An error occurred.",
            description: "Unable to login to your account.",
            status: "error",
            duration: 9000,
            isClosable: true
          });
        });
    }
  });

  useEffect(() => {
    console.log("running");
  });

  const [answers, setAnswers] = useState(emptyAnswers);

  const handleChange = e => {
    setAnswers({
      ...answers,
      [e.target.name]: !(e.target.value === "")
        ? e.target.value
        : e.target.innerText
    });
  };
  console.log(answers);

  const handleSubmit = async e => {
    // try {
    //   e.preventDefault();
    //   const res = await client.query({
    //     query: GET_UNIT
    //   });
    //   console.log(res.data);
    //   debugger;
    // } catch (err) {
    //   console.log(err);
    //   debugger;
    // }
    try {
      console.log(answers.heightUnit);
      e.preventDefault();
      const res = await client.mutate({
        mutation: ONBOARDING,
        variables: {
          id: userData.user_id,
          heightUnit: userData.user_id,
          weightUnit: userData.user_id,
          goal: answers.goal,
          experience: answers.experience,
          equipment: true
        }
      });
      console.log(res.data);
      toast({
        title: "Onboarding complete",
        description: "You can now access your dashboard",
        status: "success",
        duration: 9000,
        isClosable: true
      });
      history.push("/");
    } catch (err) {
      console.log(err);
      toast({
        title: "Unable to complete onboarding",
        description: "Kindly check input and try again.",
        status: "error",
        duration: 9000,
        isClosable: true
      });
    }
  };

  const CustomRadio = React.forwardRef((props, ref) => {
    const { isChecked, isDisabled, value, onSubmit, ...rest } = props;
    return (
      <Button
        ref={ref}
        color={isChecked ? "#ff8744" : "#CCC5B9"}
        borderColor={isChecked ? "#ff8744" : "#CCC5B9"}
        aria-checked={isChecked}
        role="radio"
        isDisabled={isDisabled}
        {...rest}
      />
    );
  });

  return (
    <AuthStyle>
      <div className="auth-container">
        <div className="auth-banner">
          <Preview pageName="Preferences" />
        </div>

        <div className="auth-form">
          <div className="logo">
            <Logo />
          </div>

          <form onSubmit={formik.handleSubmit}>
            <h2>Preferences</h2>

            <div className="body-status">
              <p>Which weight measurement unit do you prefer?</p>
              <RadioButtonGroup
                name="heightUnit"
                className="btnGroup"
                defaultValue="kilogram"
                onClick={handleChange}
                isInline
              >
                <CustomRadio className="unitButton" value="kilogram">
                  kilogram
                </CustomRadio>
                <CustomRadio className="unitButton" value="pounds">
                  pounds
                </CustomRadio>
              </RadioButtonGroup>
              <p>Which height measurement unit do you prefer?</p>
              <RadioButtonGroup
                name="weightUnit"
                className="btnGroup"
                defaultValue="rad2"
                onClick={handleChange}
                isInline
              >
                <CustomRadio className="unitButton" value="meters">
                  meters
                </CustomRadio>
                <CustomRadio className="unitButton" value="inches">
                  inches
                </CustomRadio>
              </RadioButtonGroup>
            </div>
            <p>What is your fitness goal?</p>
            <Select
              name="goal"
              className="dropdown"
              placeholder="Select a goal"
              options={[
                { value: "Weight Loss", text: "Weight Loss" },
                { value: "Muscle Gain", text: "Muscle Gain" },
                { value: "Athletic", text: "Athletic" },
                { value: "Healthy", text: "Healthy" }
              ]}
              error={formik.errors.goal}
              value={formik.values.goal}
              onChange={formik.handleChange}
            />

            <p>How experienced are you working out</p>
            <Select
              name="experience"
              placeholder="Select a Level"
              options={[
                { value: "Beginner", text: "Beginner" },
                { value: "Intermediate", text: "Intermediate" },
                { value: "Expert", text: "Expert" }
              ]}
              error={formik.errors.experience}
              value={formik.values.experience}
              onChange={formik.handleChange}
            />

            <p>What workout equipment do you have?</p>
            <Select
              name="equipment"
              className="dropdown"
              placeholder="Select equipment"
              options={[
                { value: "false", text: "None" },
                { value: "true", text: "Gym" }
              ]}
              error={formik.errors.equipment}
              value={formik.values.equipment}
              onChange={formik.handleChange}
            />

            <Button
              type="submit"
              variantColor="orange"
              rightIcon="arrow-forward"
              className="auth-form-button"
              size="lg"
            >
              Submit
            </Button>
          </form>
        </div>
      </div>
    </AuthStyle>
  );
}

export default withApollo(Onboarding);
