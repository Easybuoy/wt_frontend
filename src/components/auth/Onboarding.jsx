import React, { useState, useEffect } from "react";
import { withApollo } from "react-apollo";
import { Button, RadioButtonGroup, useToast, Box, Flex } from "@chakra-ui/core";
import Input from "../common/Input";
import { useFormik } from "formik";
import * as yup from "yup";
import PropTypes from "prop-types";

import Preview from "../common/Preview";
import AuthStyle from "./AuthStyle";
import Select from "../common/Select";
import Logo from "../common/Logo";
import { ONBOARDING } from "../../graphql/mutations";
import { GET_UNITS } from "../../graphql/queries";
import { getUserDetails, userOnboardedSuccessfully } from "../../utils";
import { Redirect } from "react-router-dom";
import CustomSpinner from "../common/Spinner";

const userData = getUserDetails();

const Onboarding = ({ client, history }) => {
  const toast = useToast();
  const [heightUnits, setHeightUnits] = useState([]);
  const [weightUnits, setWeightUnits] = useState([]);
  const [isLoading, setIsLoading] = useState(true);

  const alert = (title, description, status) => {
    toast({
      title,
      description,
      status,
      duration: 9000,
      isClosable: true
    });
  };

  const formik = useFormik({
    initialValues: {
      goal: "",
      equipment: "",
      experience: "",
      heightUnit: "",
      weightUnit: "",
      height: "",
      weight: ""
    },
    validationSchema: yup.object().shape({
      goal: yup.string().required("Please select your workout goal"),
      equipment: yup.string().required("Please select your workout equipment"),
      experience: yup
        .string()
        .required("Please select your workout experience"),
      heightUnit: yup
        .string()
        .required("Please select your workout preferred height unit"),
      weightUnit: yup
        .string()
        .required("Please select your workout preferred weight unit"),
      weight: yup.number().required("Please enter your weight"),
      height: yup.number().required("Please enter your height")
    }),

    onSubmit: value => {
      client
        .mutate({
          mutation: ONBOARDING,
          variables: {
            id: userData.user_id,
            heightUnit: value.heightUnit,
            weightUnit: value.weightUnit,
            goal: value.goal,
            experience: value.experience,
            equipment: JSON.parse(value.equipment),
            weight: value.weight,
            height: value.height
          }
        })
        .then(res => {
          userOnboardedSuccessfully("yes");

          alert(
            "Onboarding Completed.",
            "You can now access your dashboard",
            "success"
          );
          history.push("/");
          window.location.reload();
        })
        .catch(() => {
          console.log("aaa");
          alert(
            "An error occurred...",
            "Unable to complete onboarding. Please try again",
            "error"
          );
        });
    }
  });

  useEffect(() => {
    client
      .query({
        query: GET_UNITS
      })
      .then(res => {
        const weightUnit = res.data.units.filter(
          unit => unit.type === "weight"
        );
        const heightUnit = res.data.units.filter(
          unit => unit.type === "height"
        );
        setHeightUnits(heightUnit);
        setWeightUnits(weightUnit);
        setIsLoading(false);
      })
      .catch(() => {
        setIsLoading(false);
        console.log("ab");
        alert(
          "An error occurred....",
          "Unable to complete onboarding. Please reload the page and try again",
          "error"
        );
      });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  if (userData.isNewUser === false) {
    alert("Onboarding already completed.", "Proceed to workout", "warning");
    return <Redirect to="/" />;
  }

  const CustomRadio = React.forwardRef((props, ref) => {
    const { isChecked, name, isDisabled, value, onClick, ...rest } = props;
    return (
      <>
        <Button
          ref={ref}
          name={name}
          color={isChecked ? "#ff8744" : "#CCC5B9"}
          borderColor={isChecked ? "#ff8744" : "#CCC5B9"}
          aria-checked={isChecked}
          role="radio"
          isDisabled={isDisabled}
          value={value}
          onClick={onClick}
          {...rest}
        />
      </>
    );
  });

  if (isLoading) {
    return (
      <Box>
        <Flex
          width="100vw"
          height="100vh"
          justifyContent="center"
          align="center"
        >
          <CustomSpinner thickness="6px" size="xl" text="Loading..." />
        </Flex>
      </Box>
    );
  }
  return (
    <AuthStyle>
      <div className="auth-container">
        <div className="auth-banner">
          <Preview pageName="Preferences" />
        </div>

        <div className="auth-form">
          <div className="logo">
            <Logo />
          </div>

          <form onSubmit={formik.handleSubmit}>
            <h2>Preferences</h2>

            <p>What is your height?</p>
            <Input
              name="height"
              error={formik.errors.height}
              value={formik.values.height}
              placeholder="Enter Height"
              onChange={formik.handleChange}
              type="number"
              variant="filled"
              id="height"
              bg="#fffcf2"
              focusBorderColor="#dd6b20"
              onBlur={formik.handleBlur}
              touchedName={formik.touched.height}
            />

            <div className="body-status">
              <p>Which height measurement unit do you prefer?</p>
              <RadioButtonGroup
                name="heightUnit"
                className="btnGroup"
                defaultValue="kilogram"
                onClick={formik.handleChange}
                isInline
                err={formik.errors.heightUnit}
                value={formik.values.heightUnit}
                onBlur={formik.handleBlur}
              >
                {heightUnits.map(heightUnit => (
                  <CustomRadio
                    key={heightUnit.name}
                    className="unitButton"
                    value={heightUnit.id}
                  >
                    {heightUnit.name.charAt(0).toUpperCase() +
                      heightUnit.name.slice(1)}
                  </CustomRadio>
                ))}
              </RadioButtonGroup>
              {formik.touched.heightUnit && formik.errors.heightUnit ? (
                <span>{formik.errors.heightUnit}</span>
              ) : null}

              <p>What is your weight?</p>
              <Input
                name="weight"
                error={formik.errors.weight}
                value={formik.values.weight}
                onChange={formik.handleChange}
                placeholder="Enter Weight"
                variant="filled"
                id="weight"
                type="number"
                bg="#fffcf2"
                focusBorderColor="#dd6b20"
                onBlur={formik.handleBlur}
                touchedName={formik.touched.weight}
              />

              <p>Which weight measurement unit do you prefer?</p>
              <RadioButtonGroup
                name="weightUnit"
                className="btnGroup"
                defaultValue="rad2"
                onClick={formik.handleChange}
                isInline
                value={formik.values.weightUnit}
                error={formik.errors.weightUnit}
                onBlur={formik.handleBlur}
              >
                {weightUnits.map(weightUnit => (
                  <CustomRadio
                    key={weightUnit.name}
                    className="unitButton"
                    value={weightUnit.id}
                  >
                    {weightUnit.name.charAt(0).toUpperCase() +
                      weightUnit.name.slice(1)}
                  </CustomRadio>
                ))}
              </RadioButtonGroup>
              {formik.touched.weightUnit && formik.errors.weightUnit ? (
                <span>{formik.errors.weightUnit}</span>
              ) : null}
            </div>
            <p>What is your fitness goal?</p>
            <Select
              name="goal"
              className="dropdown"
              placeholder="Select a goal"
              options={[
                { value: "Weight Loss", text: "Weight Loss" },
                { value: "Muscle Gain", text: "Muscle Gain" },
                { value: "Athletic", text: "Athletic" },
                { value: "Healthy", text: "Healthy" }
              ]}
              error={formik.errors.goal}
              value={formik.values.goal}
              onChange={formik.handleChange}
              onBlur={formik.handleBlur}
              touchedName={formik.touched.goal}
            />

            <p>How experienced are you working out</p>
            <Select
              name="experience"
              placeholder="Select a Level"
              options={[
                { value: "Beginner", text: "Beginner" },
                { value: "Intermediate", text: "Intermediate" },
                { value: "Expert", text: "Expert" }
              ]}
              error={formik.errors.experience}
              value={formik.values.experience}
              onChange={formik.handleChange}
              onBlur={formik.handleBlur}
              touchedName={formik.touched.experience}
            />

            <p>What workout equipment do you have?</p>
            <Select
              name="equipment"
              className="dropdown"
              placeholder="Select equipment"
              options={[
                { value: "false", text: "None" },
                { value: "true", text: "Gym" }
              ]}
              error={formik.errors.equipment}
              value={formik.values.equipment}
              onChange={formik.handleChange}
              onBlur={formik.handleBlur}
              touchedName={formik.touched.equipment}
            />

            <Button
              type="submit"
              variantColor="orange"
              rightIcon="arrow-forward"
              className="auth-form-button"
              size="lg"
            >
              Submit
            </Button>
          </form>
        </div>
      </div>
    </AuthStyle>
  );
};

export default withApollo(Onboarding);

Onboarding.propTypes = {
  client: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired
};
