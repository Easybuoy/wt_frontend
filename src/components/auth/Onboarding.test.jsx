import React from "react";
import { shallow } from "enzyme";

import Onboarding from "./Onboarding";

describe("<Onboarding />", () => {
  let wrapper;

  const props = {
    location: { search: jest.mock() },
    client: { mutate: jest.fn(), query: jest.fn() },
    history: { push: jest.fn() }
  };

  beforeEach(() => {
    wrapper = shallow(<Onboarding {...props} />);
  });

  it("renders the Onboarding component correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
