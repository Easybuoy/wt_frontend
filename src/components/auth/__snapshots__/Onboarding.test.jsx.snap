// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Onboarding /> renders the Onboarding component correctly 1`] = `
<ApolloConsumer>
  <Component />
</ApolloConsumer>
`;
